Appliance postgres selabel

kbrock opened the pull request:
The selinux labels have been fixed for the collections version of postgres.
We needed to remove one of the labels, thinking we should just remove the second.

1. After mounting the postgres volume, relabel all of the postgres, not just the log directory.
2. ~~Setup the selinux labels to match the suggested install.~~

~~@jrafanie do you remember why we have the selinux label on `pg_log`? Is this to support log rotation or something else?~~ answer: Yes, for `logrotate`

[skip ci]

**UPDATE:** not changing label, just changing directory that is relabeled (will change log directory later)

jrafanie commented:
@kbrock You're right, log rotation.  The linked bz below is "postgresql server log is not being processed by logrotate"

```
commit 3aedacce061d6e9080bc401bbc74bf9d3b45b6e8
Author: Steve Linabery <slinaber@redhat.com>
Date:   Tue Sep 17 16:22:48 2013 -0500

    BZ 1009084 Apply new selinux label to postgresql log directory

    https://bugzilla.redhat.com/show_bug.cgi?id=1009084

    Change the label on the postgresql log directory to var_log_t,
    and apply the label after the database disk and postgresql database
    are initialized.
```

Is this still needed?  I'd assume so or selinux might fail.  You'd have to run an appliance with and without the line and force logrotate to run to make sure.

kbrock commented:
@jrafanie My reasoning:
we are able to rotate the standard postgresql logs.
Now that collections has the correct labels, we should be able to rotate the logs with default labels.

I'll take it for a spin.

any thoughts on the relabel?

jrafanie commented on system/cfme-setup.sh:
```diff
@@ -40,9 +40,6 @@ EOF
 
-# relabel the pg_log directory in postgresql datadir, but defer restorecon
-# until after the database is initialized during firstboot configuration
-/usr/sbin/semanage fcontext -a -t var_log_t "/opt/rh/postgresql92/root/var/lib/pgsql/data/pg_log(/.*)?"
```
@kbrock I don't understand the details on what this `semanage` WAS doing vs. what the change you made to do `restorecon` on the whole DATABASE_DISK_MOUNT_POINT vs. just the log directory.

kbrock replied:
This is 2 PRs around 1 central theme.

1. When you mount `DATABASE_DISK_MOUNT_POINT` setup all of it to have the proper privileges.
2. Just use the default fcontext for all the directories in `DATABASE_DISK_MOUNT_POINT`

It is possible that task 2 is wrong. If that is the case I'll remove the change to `cfme-setup.sh`.

jrafanie replied:
Is this something you can test on a nightly appliances before merging?  I'm fine with changing it but we need to ensure the db configuration works and the logrotate of pg logs works.

kbrock commented:
@jrafanie Goal is still to change the label on that directory.
But for now, am happy relabeling a directory up.

Since we didn't change the rules, it shouldn't change anything.

jrafanie commented:
@kbrock Can you remind me the "why"?  I don't really understand.  Are we getting selinux audit failures?

Also, was this tested with an internal db setup as part of a centos appliance?  The postgresql setup is very order dependent so I'm cautious about any changes there without tests..

kbrock commented:
@jrafanie why:

scl postgres had no policies around postgres database.
We had to add scl policy around these files. (as a crutch)

scl postgres added policies around postgres databases
This broke our policies so we removed 1/2 of our policies.
Since scl postgres fixed their policies, it seems we wouldn't need any around these files.

Also, we don't know what policies postgres wants to add to the rest of the database directories. So if we are applying privileges, lets do it to the whole directory we just added rather than just the logs subdirectory.

Originally the change I suggested here was to remove the remaining postgres policy, but I backed off and am suggesting a more timid one. Relabel the whole postgres directory. We (could just relabel the whole operating system.)

kbrock commented:
@jrafanie This just adds one more layer to apply privileges. Reasoning is to start relabeling from the directory we mounted. This PR does not change any labels.

---
personal notes:
Going back to submit a pr in postgres and scl:

1. Add privileges set on pg_log

    semanage fcontext -a -t var_postgresql_log `/var/lib/pgsql/data/pg_log(/.*)?'
2. logrotate supports `postgresql_log_t` in addition to currently supported `var_log_t`
3. add equivalency for data

    semanage fcontext -a -e / /opt/rh/postgresql92/root

useful commands (not useful enough)
```
run-parts /etc/cron.daily -v
sandbox -C -t logrotate_t /usr/sbin/logrotate -d /etc/logrotate.conf 
```

jrafanie commented:
> Add privileges set on pg_log
> 
> semanage fcontext -a -t var_postgresql_log/var/lib/pgsql/data/pg_log(/.*)?'`

What does that mean?   How does that differ from now and how will that benefit us?

> logrotate supports postgresql_log_t in addition to currently supported var_log_t

Again, what does that mean?  Is it better to specify more specific context, `postgrsql_log_t`?  What's our benefit?  My thinking is as long as selinux doesn't prevent our stuff from working, I'm good... what does it improve to use this?
 
> add equivalency for data
> 
> semanage fcontext -a -e / /opt/rh/postgresql92/root

I don't know what that means.


As long as you test a end-to-end appliance build with this setup, I'm fine with tweaking some selinux knobs but I'm cautious about making large changes since I don't understand selinux.  If you understand it and can explain it to others, I'm fine with changes to be make us more "standard" or eliminate future problems.

kbrock commented:
@jrafanie In this pr:
1. no selinux label changes
2. yes relabel whole directory that was just mounted rather than a just a subset.

Those are personal notes and for future reference. 7.1 will probably have different selinux privileges around postgres anyway. I want to capture my findings so I can work with the rpm developers of scl postgres and possibly logrotate to fix their privileges.

I researched and learned some stuff. I hit a brick wall. I want to write down those changes so I can access them again in the future.

Fryguy commented:
Bump.

jrafanie commented:
@kbrock Should we merge this for the sprint build?  I don't understand these changes to know if there's any risk of it breaking the appliance build.

kbrock commented:
- upstream will not notice since it is not mounting an external disk for the db.
- downstream should not notice any difference for the disk privileges. but this is targeted there.

It ensures the whole database disk mount gets labeled, not just a subset.
